Assignment:
Use the dataset “Titanic“ as Pandas Data Frame. Implement a Regression model with multiple variables: independent variables “Age” and “Fare” and “Survived“ as dependent variable.

In [2]:
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline
from pandas.core import datetools

/Users/mariiashcherbiak/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [3]:

df = pd.read_csv('/Users/mariiashcherbiak/Downloads/BTS_MasterInBigData-master-10/Session_3_CDA/FileData/1_titanic_dataset.csv', index_col=0)
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

Regression model with two variables and with a constant

In [4]:
X = df[["Age", "Fare"]]#independent variables
y = df["Survived"]#dependent variable

Replace mising values

In [5]:
count_nan=len(df)-df.count()
count_nan

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [6]:
y.head()

PassengerId
1    0
2    1
3    1
4    1
5    0
Name: Survived, dtype: int64

In [7]:
df["Age"].isnull().sum()

177

In [8]:
mean=df["Age"].mean()
mean

29.69911764705882

In [9]:
df["Age"] = df["Age"].fillna(29)

In [10]:
df["Cabin"] = df["Cabin"].fillna("G")

In [11]:
df["Embarked"] = df["Embarked"].fillna("S")

In [12]:
count_nan=len(df)-df.count()
count_nan

Survived    0
Pclass      0
Name        0
Sex         0
Age         0
SibSp       0
Parch       0
Ticket      0
Fare        0
Cabin       0
Embarked    0
dtype: int64

In [13]:
X = df[["Age", "Fare"]]#independent variables
y = df["Survived"]#dependent variable

In [14]:
X = sm.add_constant(X) #Adds a constant term to the predictor
X.head()

const   Age     Fare
PassengerId                      
1              1.0  22.0   7.2500
2              1.0  38.0  71.2833
3              1.0  26.0   7.9250
4              1.0  35.0  53.1000
5              1.0  35.0   8.0500

In [15]:
model = sm.OLS(y, X).fit()#Argument order of the Regression mode

In [16]:
predictions = model.predict(X) #Predictions by the mode
predictions

PassengerId
1      0.345076
2      0.456339
3      0.332954
4      0.419388
5      0.302051
6      0.323934
7      0.350237
8      0.450483
9      0.337841
10     0.432274
11     0.432148
12     0.270428
13     0.354099
14     0.348663
15     0.374409
16     0.253359
17     0.471450
18     0.335763
19     0.341846
20     0.320721
21     0.348803
22     0.318413
23     0.371395
24     0.397837
25     0.429663
26     0.352426
27     0.320721
28     1.021615
29     0.322425
30     0.322468
         ...   
862    0.359615
863    0.303510
864    0.483054
865    0.353112
866    0.290654
867    0.344938
868    0.426485
869    0.326647
870    0.417649
871    0.332878
872    0.376328
873    0.301046
874    0.262886
875    0.367884
876    0.369300
877    0.358777
878    0.357168
879    0.322468
880    0.424810
881    0.383502
882    0.308589
883    0.353584
884    0.332721
885    0.334145
886    0.343063
887    0.342703
888    0.414740
889    0.362981
890    0.390451
891    0.311679
Length: 891,

In [17]:
model.summary()# statistics

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               Survived   R-squared:                       0.075
Model:                            OLS   Adj. R-squared:                  0.073
Method:                 Least Squares   F-statistic:                     35.86
Date:                Fri, 10 Nov 2017   Prob (F-statistic):           1.05e-15
Time:                        22:25:57   Log-Likelihood:                -587.36
No. Observations:                 891   AIC:                             1181.
Df Residuals:                     888   BIC:                             1195.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.4025      0.040     10.165      0.000       0.325       0.480
Age           -0.0035      0.001     -2.860      0.004      -0.006      -0.001
Fare           0.0026      0.000      8.204      0.000       0.002       0.003
==============================================================================
Omnibus:                       31.284   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              121.166
Skew:                           0.479   Prob(JB):                     4.89e-27
Kurtosis:                       1.468   Cond. No.                         157.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R2 of model with a constant:0.075. Very bad model

Regression model with two variables and without a constant:

In [18]:
X = df[["Age", "Fare"]]#independent variables
y = df["Survived"]#dependent variable

In [88]:
X = sm.add_constant(X) 
X.head()

const   Age     Fare
PassengerId                      
1              1.0  22.0   7.2500
2              1.0  38.0  71.2833
3              1.0  26.0   7.9250
4              1.0  35.0  53.1000
5              1.0  35.0   8.0500

In [89]:
X["const"] = X["const"].apply(lambda x: x*0)#Removing the constant term to the predictor.
X

const   Age      Fare
PassengerId                       
1              0.0  22.0    7.2500
2              0.0  38.0   71.2833
3              0.0  26.0    7.9250
4              0.0  35.0   53.1000
5              0.0  35.0    8.0500
6              0.0  29.0    8.4583
7              0.0  54.0   51.8625
8              0.0   2.0   21.0750
9              0.0  27.0   11.1333
10             0.0  14.0   30.0708
11             0.0   4.0   16.7000
12             0.0  58.0   26.5500
13             0.0  20.0    8.0500
14             0.0  39.0   31.2750
15             0.0  14.0    7.8542
16             0.0  55.0   16.0000
17             0.0   2.0   29.1250
18             0.0  29.0   13.0000
19             0.0  31.0   18.0000
20             0.0  29.0    7.2250
21             0.0  35.0   26.0000
22             0.0  34.0   13.0000
23             0.0  15.0    8.0292
24             0.0  28.0   35.5000
25             0.0   8.0   21.0750
26             0.0  38.0   31.3875
27             0.0  29.0    7.2250
28             0.0  19.0  263.0000
29             0.0  29.0    7.8792
30             0.0  29.0    7.8958
...            ...   ...       ...
862            0.0  21.0   11.5000
863            0.0  48.0   25.9292
864            0.0  29.0   69.5500
865            0.0  24.0   13.0000
866            0.0  42.0   13.0000
867            0.0  27.0   13.8583
868            0.0  31.0   50.4958
869            0.0  29.0    9.5000
870            0.0   4.0   11.1333
871            0.0  26.0    7.8958
872            0.0  47.0   52.5542
873            0.0  33.0    5.0000
874            0.0  47.0    9.0000
875            0.0  28.0   24.0000
876            0.0  15.0    7.2250
877            0.0  20.0    9.8458
878            0.0  19.0    7.8958
879            0.0  29.0    7.8958
880            0.0  56.0   83.1583
881            0.0  25.0   26.0000
882            0.0  33.0    7.8958
883            0.0  22.0   10.5167
884            0.0  28.0   10.5000
885            0.0  25.0    7.0500
886            0.0  39.0   29.1250
887            0.0  27.0   13.0000
888            0.0  19.0   30.0000
889            0.0  29.0   23.4500
890            0.0  26.0   30.0000
891            0.0  32.0    7.7500

[891 rows x 3 columns]

In [90]:
model = sm.OLS(y, X).fit()

In [91]:
predictions = model.predict(X) 
predictions

PassengerId
1      0.185697
2      0.506664
3      0.217425
4      0.426939
5      0.284403
6      0.241306
7      0.563587
8      0.081477
9      0.234974
10     0.198716
11     0.082430
12     0.513092
13     0.173432
14     0.387478
15     0.128424
16     0.457518
17     0.106946
18     0.255676
19     0.286292
20     0.237404
21     0.341196
22     0.292666
23     0.136375
24     0.319467
25     0.125865
26     0.380436
27     0.237404
28     0.972684
29     0.239474
30     0.239526
         ...   
862    0.191745
863    0.437147
864    0.434597
865    0.218685
866    0.351851
867    0.243595
868    0.389107
869    0.244602
870    0.064818
871    0.217332
872    0.513989
873    0.259957
874    0.376186
875    0.283081
876    0.133831
877    0.179113
878    0.165546
879    0.239526
880    0.677402
881    0.267215
882    0.269119
883    0.196032
884    0.240368
885    0.207258
886    0.380675
887    0.240880
888    0.235482
889    0.288739
890    0.287269
891    0.261259
Length: 891,

In [93]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               Survived   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.362
Method:                 Least Squares   F-statistic:                     253.9
Date:                Fri, 10 Nov 2017   Prob (F-statistic):           5.94e-88
Time:                        13:23:18   Log-Likelihood:                -636.40
No. Observations:                 891   AIC:                             1277.
Df Residuals:                     889   BIC:                             1286.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const               0          0        nan        nan           0           0
Age            0.0074      0.001     12.224      0.000       0.006       0.009
Fare           0.0032      0.000      9.583      0.000       0.003       0.004
==============================================================================
Omnibus:                      970.601   Durbin-Watson:                   1.939
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               87.362
Skew:                           0.449   Prob(JB):                     1.07e-19
Kurtosis:                       1.756   Cond. No.                          inf
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

R2 in model without a constant higther than in the previous model: 0.364